{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Which annotation files to use for FC_directed?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": "/Users/davidebuzzao/Projects/funcoup_directed\n"
    }
   ],
   "source": [
    "%cd ~/projects/funcoup_directed/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd, pickle\n",
    "genome = 'GRCh38'\n",
    "organism = 'human'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ENCODE\n",
    "\n",
    "The human genome is estimated to encode approximately 2000 TFs, which operate programs that change cellular states by binding to proxy or distal cis-regulatory elements (CREs) for a set of target genes. Interactions between TFs and CREs of target genes are generally detected by DNA-binding experiments such as chromatin immunoprecipitation (ChIP), which is often followed by microarray analysis (ChIP-chip) or deep-sequencing analysis (ChIP-seq). \n",
    "ChIP-seq is therefore one of the methods used to analyze protein interactions with DNA. It combines chromatin immunoprecipitation with DNA sequencing to infer the possible binding sites of DNA-associated proteins. ChIP-seq data are available on the Encyclopedia of DNA Elements (ENCODE) portal (Davis et al., 2018) and can be used to link transcription factors with their targets. The ENCODE project, in fact, has been formed by the NHGRI in 2003 with the aim of bringing together a variety of experimental and computational labs to identify biochemically active regions in the human and mouse genomes utilizing a variety of HT approaches. Multiple pipelines has been developed by the ENCODE consortium, a couple of which (in particular) to study two different classes of protein-chromatin interactions: \n",
    "1. the pipeline for transcription factor ChIP-seq (TF ChIP-seq) \n",
    "2. the pipeline for histone ChIP-seq. \n",
    "\n",
    "Both pipelines are maintained in the ENCODE Data Coordinating Center (DCC) GitHub repository (https://github.com/ENCODE-DCC/chip-seq-pipeline2).  \n",
    "\n",
    "For what concerns the transcription factor ChIP-seq (TF ChIP-seq) pipeline, it is suitable for proteins that are expected to bind in a punctate manner, such as to specific DNA sequences or specific chromatin configurations. In these assays, the IP target is typically a known or putative transcription factor or chromatin remodeller but can also be an RNA-binding protein or other DNA- or chromatin-specific factor. For the FunCoup purpose, everything but transcription factors’ assays is filtered out, remaining with 2715 TF-ChIP-seq experiments on Homo sapiens, Drosophila melanogaster, Caenorhabditis elegans and Mus musculus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Line magic function `%wget` not found.\n"
     ]
    }
   ],
   "source": [
    "%wget -O data/evidences/encode/human/metadata.tsv $(head -n +1 data/evidences/encode/human/files.txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": "      File accession        File format File type File format type  \\\n0        ENCFF827IYA                bam       bam              NaN   \n1        ENCFF393IBD                bam       bam              NaN   \n2        ENCFF123GSL             bigWig    bigWig              NaN   \n3        ENCFF174SPG     bed narrowPeak       bed       narrowPeak   \n4        ENCFF198ATT  bigBed narrowPeak    bigBed       narrowPeak   \n...              ...                ...       ...              ...   \n60602    ENCFF724MHH  bigBed narrowPeak    bigBed       narrowPeak   \n60603    ENCFF722GFS     bed narrowPeak       bed       narrowPeak   \n60604    ENCFF894NYR  bigBed narrowPeak    bigBed       narrowPeak   \n60605    ENCFF506PCG  bigBed narrowPeak    bigBed       narrowPeak   \n60606    ENCFF553PWA     bed narrowPeak       bed       narrowPeak   \n\n                                 Output type Experiment accession     Assay  \\\n0                      unfiltered alignments          ENCSR445ACU  ChIP-seq   \n1                      unfiltered alignments          ENCSR445ACU  ChIP-seq   \n2                             signal p-value          ENCSR445ACU  ChIP-seq   \n3      peaks and background as input for IDR          ENCSR445ACU  ChIP-seq   \n4      peaks and background as input for IDR          ENCSR445ACU  ChIP-seq   \n...                                      ...                  ...       ...   \n60602  peaks and background as input for IDR          ENCSR000BLE  ChIP-seq   \n60603     conservative IDR thresholded peaks          ENCSR000BLE  ChIP-seq   \n60604          optimal IDR thresholded peaks          ENCSR000BLE  ChIP-seq   \n60605     conservative IDR thresholded peaks          ENCSR000BLE  ChIP-seq   \n60606          optimal IDR thresholded peaks          ENCSR000BLE  ChIP-seq   \n\n      Biosample term id Biosample term name Biosample type  ... Assembly  \\\n0           EFO:0001187               HepG2      cell line  ...     hg19   \n1           EFO:0001187               HepG2      cell line  ...     hg19   \n2           EFO:0001187               HepG2      cell line  ...     hg19   \n3           EFO:0001187               HepG2      cell line  ...     hg19   \n4           EFO:0001187               HepG2      cell line  ...     hg19   \n...                 ...                 ...            ...  ...      ...   \n60602       EFO:0001187               HepG2      cell line  ...     hg19   \n60603       EFO:0001187               HepG2      cell line  ...     hg19   \n60604       EFO:0001187               HepG2      cell line  ...     hg19   \n60605       EFO:0001187               HepG2      cell line  ...     hg19   \n60606       EFO:0001187               HepG2      cell line  ...     hg19   \n\n      Genome annotation Platform Controlled by File Status  \\\n0                   NaN      NaN           NaN    released   \n1                   NaN      NaN           NaN    released   \n2                   NaN      NaN           NaN    released   \n3                   NaN      NaN           NaN    released   \n4                   NaN      NaN           NaN    released   \n...                 ...      ...           ...         ...   \n60602               NaN      NaN           NaN    released   \n60603               NaN      NaN           NaN    released   \n60604               NaN      NaN           NaN    released   \n60605               NaN      NaN           NaN    released   \n60606               NaN      NaN           NaN    released   \n\n                                                  s3_uri Audit WARNING  \\\n0      s3://encode-public/2018/02/21/e515ec26-4792-40...           NaN   \n1      s3://encode-public/2018/02/21/924de693-bf20-4d...           NaN   \n2      s3://encode-public/2018/02/26/48e60a16-f62a-4f...           NaN   \n3      s3://encode-public/2018/02/26/4445d47e-1624-4b...           NaN   \n4      s3://encode-public/2018/02/26/5a89254e-fec5-4e...           NaN   \n...                                                  ...           ...   \n60602  s3://encode-public/2017/07/07/422dcd55-6fff-43...           NaN   \n60603  s3://encode-public/2017/07/07/eb2b9798-4787-4d...           NaN   \n60604  s3://encode-public/2017/07/07/a7bb1f7a-4bd3-41...           NaN   \n60605  s3://encode-public/2017/07/07/51d513dd-7bf6-4b...           NaN   \n60606  s3://encode-public/2017/07/07/e64833ca-eaaf-4f...           NaN   \n\n       Audit INTERNAL_ACTION Audit NOT_COMPLIANT  Audit ERROR  \n0                        NaN                 NaN          NaN  \n1                        NaN                 NaN          NaN  \n2                        NaN                 NaN          NaN  \n3                        NaN                 NaN          NaN  \n4                        NaN                 NaN          NaN  \n...                      ...                 ...          ...  \n60602                    NaN                 NaN          NaN  \n60603                    NaN                 NaN          NaN  \n60604                    NaN                 NaN          NaN  \n60605                    NaN                 NaN          NaN  \n60606                    NaN                 NaN          NaN  \n\n[60607 rows x 55 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>File accession</th>\n      <th>File format</th>\n      <th>File type</th>\n      <th>File format type</th>\n      <th>Output type</th>\n      <th>Experiment accession</th>\n      <th>Assay</th>\n      <th>Biosample term id</th>\n      <th>Biosample term name</th>\n      <th>Biosample type</th>\n      <th>...</th>\n      <th>Assembly</th>\n      <th>Genome annotation</th>\n      <th>Platform</th>\n      <th>Controlled by</th>\n      <th>File Status</th>\n      <th>s3_uri</th>\n      <th>Audit WARNING</th>\n      <th>Audit INTERNAL_ACTION</th>\n      <th>Audit NOT_COMPLIANT</th>\n      <th>Audit ERROR</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ENCFF827IYA</td>\n      <td>bam</td>\n      <td>bam</td>\n      <td>NaN</td>\n      <td>unfiltered alignments</td>\n      <td>ENCSR445ACU</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2018/02/21/e515ec26-4792-40...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ENCFF393IBD</td>\n      <td>bam</td>\n      <td>bam</td>\n      <td>NaN</td>\n      <td>unfiltered alignments</td>\n      <td>ENCSR445ACU</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2018/02/21/924de693-bf20-4d...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENCFF123GSL</td>\n      <td>bigWig</td>\n      <td>bigWig</td>\n      <td>NaN</td>\n      <td>signal p-value</td>\n      <td>ENCSR445ACU</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2018/02/26/48e60a16-f62a-4f...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENCFF174SPG</td>\n      <td>bed narrowPeak</td>\n      <td>bed</td>\n      <td>narrowPeak</td>\n      <td>peaks and background as input for IDR</td>\n      <td>ENCSR445ACU</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2018/02/26/4445d47e-1624-4b...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENCFF198ATT</td>\n      <td>bigBed narrowPeak</td>\n      <td>bigBed</td>\n      <td>narrowPeak</td>\n      <td>peaks and background as input for IDR</td>\n      <td>ENCSR445ACU</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2018/02/26/5a89254e-fec5-4e...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>60602</th>\n      <td>ENCFF724MHH</td>\n      <td>bigBed narrowPeak</td>\n      <td>bigBed</td>\n      <td>narrowPeak</td>\n      <td>peaks and background as input for IDR</td>\n      <td>ENCSR000BLE</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2017/07/07/422dcd55-6fff-43...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>60603</th>\n      <td>ENCFF722GFS</td>\n      <td>bed narrowPeak</td>\n      <td>bed</td>\n      <td>narrowPeak</td>\n      <td>conservative IDR thresholded peaks</td>\n      <td>ENCSR000BLE</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2017/07/07/eb2b9798-4787-4d...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>60604</th>\n      <td>ENCFF894NYR</td>\n      <td>bigBed narrowPeak</td>\n      <td>bigBed</td>\n      <td>narrowPeak</td>\n      <td>optimal IDR thresholded peaks</td>\n      <td>ENCSR000BLE</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2017/07/07/a7bb1f7a-4bd3-41...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>60605</th>\n      <td>ENCFF506PCG</td>\n      <td>bigBed narrowPeak</td>\n      <td>bigBed</td>\n      <td>narrowPeak</td>\n      <td>conservative IDR thresholded peaks</td>\n      <td>ENCSR000BLE</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2017/07/07/51d513dd-7bf6-4b...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>60606</th>\n      <td>ENCFF553PWA</td>\n      <td>bed narrowPeak</td>\n      <td>bed</td>\n      <td>narrowPeak</td>\n      <td>optimal IDR thresholded peaks</td>\n      <td>ENCSR000BLE</td>\n      <td>ChIP-seq</td>\n      <td>EFO:0001187</td>\n      <td>HepG2</td>\n      <td>cell line</td>\n      <td>...</td>\n      <td>hg19</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>released</td>\n      <td>s3://encode-public/2017/07/07/e64833ca-eaaf-4f...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n<p>60607 rows × 55 columns</p>\n</div>"
     },
     "metadata": {},
     "execution_count": 5
    }
   ],
   "source": [
    "metadata = pd.read_csv(\"data/evidences/encode/human/metadata.tsv\", sep=\"\\t\")\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The pipeline takes both ChIP-seq reads (from paired-end, stranded or single-end, unstranded libraries) and a set of reference fastas as inputs, and outputs unfiltered and filtered alignment files in bam format (from sam: the Sequence Alignment/Mapping format is a text-based format for storing read alignments against reference sequences and it is interconvertible with the binary bam format) from the appropriate genome assembly. When multiple fastqs are generated from a single biological replicate (in multiple sequencing runs, for example), they are concatenated before mapping. With no exception: the read length should be a minimum of 50 base pairs; the sequencing platform used should be indicated and replicates should match in terms of read length and run type. Pipeline files are mapped to either the GRCh38 or mm10 sequences. Also, data that were originally collected and mapped in a non-standard fashion to earlier assemblies in prior project phases, has now been reprocessed using the ENCODE unified processing pipelines both onto a consistent hg19 (GRCh37) reference genome assembly and the updated GRCh38 reference. Similarly, mouse data previously mapped to mm9 have also been reprocessed onto the latest assembly, mm10. Reference files used in the ENCODE processing pipeline are retrievable at https://www.encodeproject.org/references/ENCSR425FOI/.\n",
    "\n",
    "Metadata in ENCODE is captured in JavaScript Object Notation (JSON). Metadata tables for datasets can be either downloaded in (.tsv) format via the web interface or obtained programmatically through the REST API (https://www.encodeproject.org/help/rest-api/). Here the subset of the information I consider necessary to be extracted from metadata and the way to retrieve them.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n +1 data/evidences/encode/human/metadata.tsv | cut -f 1,2,5,6,7,11,21,41,46 > data/evidences/encode/human/metadata_bam.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grep bam data/evidences/encode/human/metadata.tsv | grep -v unfiltered | grep '\\bGRCh38\\b' | cut -f 1,2,5,6,7,11,21,41,46 >> data/evidences/encode/human/metadata_bam.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File accession</th>\n",
       "      <th>File format</th>\n",
       "      <th>Output type</th>\n",
       "      <th>Experiment accession</th>\n",
       "      <th>Assay</th>\n",
       "      <th>Biosample organism</th>\n",
       "      <th>Experiment target</th>\n",
       "      <th>Size</th>\n",
       "      <th>Assembly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCFF908SPX</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR445ACU</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SOX13-human</td>\n",
       "      <td>1106836211</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCFF010EFV</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR445ACU</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SOX13-human</td>\n",
       "      <td>1021640457</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCFF582CJF</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR000BGM</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>USF1-human</td>\n",
       "      <td>637033775</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCFF270TGL</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR000BGM</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>USF1-human</td>\n",
       "      <td>999909157</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCFF803MUP</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR634OAQ</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CTCF-human</td>\n",
       "      <td>4302449535</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2976</th>\n",
       "      <td>ENCFF686DDV</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR000DKR</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CTCF-human</td>\n",
       "      <td>582699200</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2977</th>\n",
       "      <td>ENCFF779BHG</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR000DKR</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CTCF-human</td>\n",
       "      <td>1019590180</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2978</th>\n",
       "      <td>ENCFF205XEH</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR000AUF</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>CTCF-human</td>\n",
       "      <td>1292831532</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2979</th>\n",
       "      <td>ENCFF172IMH</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR000BLE</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FOXA1-human</td>\n",
       "      <td>497461256</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2980</th>\n",
       "      <td>ENCFF563UBP</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>ENCSR000BLE</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FOXA1-human</td>\n",
       "      <td>621267194</td>\n",
       "      <td>GRCh38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2981 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     File accession File format Output type Experiment accession     Assay  \\\n",
       "0       ENCFF908SPX         bam  alignments          ENCSR445ACU  ChIP-seq   \n",
       "1       ENCFF010EFV         bam  alignments          ENCSR445ACU  ChIP-seq   \n",
       "2       ENCFF582CJF         bam  alignments          ENCSR000BGM  ChIP-seq   \n",
       "3       ENCFF270TGL         bam  alignments          ENCSR000BGM  ChIP-seq   \n",
       "4       ENCFF803MUP         bam  alignments          ENCSR634OAQ  ChIP-seq   \n",
       "...             ...         ...         ...                  ...       ...   \n",
       "2976    ENCFF686DDV         bam  alignments          ENCSR000DKR  ChIP-seq   \n",
       "2977    ENCFF779BHG         bam  alignments          ENCSR000DKR  ChIP-seq   \n",
       "2978    ENCFF205XEH         bam  alignments          ENCSR000AUF  ChIP-seq   \n",
       "2979    ENCFF172IMH         bam  alignments          ENCSR000BLE  ChIP-seq   \n",
       "2980    ENCFF563UBP         bam  alignments          ENCSR000BLE  ChIP-seq   \n",
       "\n",
       "     Biosample organism Experiment target        Size Assembly  \n",
       "0          Homo sapiens       SOX13-human  1106836211   GRCh38  \n",
       "1          Homo sapiens       SOX13-human  1021640457   GRCh38  \n",
       "2          Homo sapiens        USF1-human   637033775   GRCh38  \n",
       "3          Homo sapiens        USF1-human   999909157   GRCh38  \n",
       "4          Homo sapiens        CTCF-human  4302449535   GRCh38  \n",
       "...                 ...               ...         ...      ...  \n",
       "2976       Homo sapiens        CTCF-human   582699200   GRCh38  \n",
       "2977       Homo sapiens        CTCF-human  1019590180   GRCh38  \n",
       "2978       Homo sapiens        CTCF-human  1292831532   GRCh38  \n",
       "2979       Homo sapiens       FOXA1-human   497461256   GRCh38  \n",
       "2980       Homo sapiens       FOXA1-human   621267194   GRCh38  \n",
       "\n",
       "[2981 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "met_bam = pd.read_csv(\"data/evidences/encode/human/metadata_bam.tsv\", sep=\"\\t\")\n",
    "met_bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the FunCoup purpose, I checked the frequency of TFs in H. sapiens in the retrievable ChIP-seq experiments. No TF is studied as much as CTCF. This is probably due to the fact the transcriptional repressor is involved in many cellular processes (including transcriptional regulation, insulator activity, V(D)J recombination and regulation of chromatin architecture) and therefore the protein could be subjected to various studies by laboratories working in different fields. The TFs that have been under investigation only once are 275/489 in human and 10/47 in mouse. \n",
    "\n",
    "Here I extract unique `Experiment accession` and exclude replication of same ChIP-Seq experiment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tail -n +2  data/evidences/encode/human/metadata_bam.tsv | cut -f 4 | uniq > data/evidences/encode/human/exp_accession.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n +1 data/evidences/encode/human/metadata_bam.tsv > data/evidences/encode/human/metadata_bam.no_rep.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in `cat data/evidences/encode/human/exp_accession.id`; do grep -m 1 $i data/evidences/encode/human/metadata_bam.tsv; done >> data/evidences/encode/human/metadata_bam.no_rep.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": "      ENCFF908SPX  bam           alignments  ENCSR445ACU  ChIP-seq  \\\n0     ENCFF582CJF  bam           alignments  ENCSR000BGM  ChIP-seq   \n1     ENCFF803MUP  bam           alignments  ENCSR634OAQ  ChIP-seq   \n2     ENCFF654KYF  bam           alignments  ENCSR000DQD  ChIP-seq   \n3     ENCFF472PUP  bam           alignments  ENCSR000DKN  ChIP-seq   \n4     ENCFF449ZQR  bam           alignments  ENCSR094ZCF  ChIP-seq   \n...           ...  ...                  ...          ...       ...   \n1513  ENCFF949KAW  bam  redacted alignments  ENCSR336PTS  ChIP-seq   \n1514  ENCFF893HMO  bam           alignments  ENCSR000DXQ  ChIP-seq   \n1515  ENCFF686DDV  bam           alignments  ENCSR000DKR  ChIP-seq   \n1516  ENCFF205XEH  bam           alignments  ENCSR000AUF  ChIP-seq   \n1517  ENCFF172IMH  bam           alignments  ENCSR000BLE  ChIP-seq   \n\n      Homo sapiens  SOX13-human  1106836211  GRCh38  \n0     Homo sapiens   USF1-human   637033775  GRCh38  \n1     Homo sapiens   CTCF-human  4302449535  GRCh38  \n2     Homo sapiens   CTCF-human   819184321  GRCh38  \n3     Homo sapiens   CTCF-human   879838417  GRCh38  \n4     Homo sapiens  CEBPG-human  1268211309  GRCh38  \n...            ...          ...         ...     ...  \n1513  Homo sapiens   CTCF-human   650522945  GRCh38  \n1514  Homo sapiens   CTCF-human   365418600  GRCh38  \n1515  Homo sapiens   CTCF-human   582699200  GRCh38  \n1516  Homo sapiens   CTCF-human  1292831532  GRCh38  \n1517  Homo sapiens  FOXA1-human   497461256  GRCh38  \n\n[1518 rows x 9 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>ENCFF908SPX</th>\n      <th>bam</th>\n      <th>alignments</th>\n      <th>ENCSR445ACU</th>\n      <th>ChIP-seq</th>\n      <th>Homo sapiens</th>\n      <th>SOX13-human</th>\n      <th>1106836211</th>\n      <th>GRCh38</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ENCFF582CJF</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR000BGM</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>USF1-human</td>\n      <td>637033775</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ENCFF803MUP</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR634OAQ</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CTCF-human</td>\n      <td>4302449535</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ENCFF654KYF</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR000DQD</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CTCF-human</td>\n      <td>819184321</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ENCFF472PUP</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR000DKN</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CTCF-human</td>\n      <td>879838417</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ENCFF449ZQR</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR094ZCF</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CEBPG-human</td>\n      <td>1268211309</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>1513</th>\n      <td>ENCFF949KAW</td>\n      <td>bam</td>\n      <td>redacted alignments</td>\n      <td>ENCSR336PTS</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CTCF-human</td>\n      <td>650522945</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>1514</th>\n      <td>ENCFF893HMO</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR000DXQ</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CTCF-human</td>\n      <td>365418600</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>1515</th>\n      <td>ENCFF686DDV</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR000DKR</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CTCF-human</td>\n      <td>582699200</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>1516</th>\n      <td>ENCFF205XEH</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR000AUF</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>CTCF-human</td>\n      <td>1292831532</td>\n      <td>GRCh38</td>\n    </tr>\n    <tr>\n      <th>1517</th>\n      <td>ENCFF172IMH</td>\n      <td>bam</td>\n      <td>alignments</td>\n      <td>ENCSR000BLE</td>\n      <td>ChIP-seq</td>\n      <td>Homo sapiens</td>\n      <td>FOXA1-human</td>\n      <td>497461256</td>\n      <td>GRCh38</td>\n    </tr>\n  </tbody>\n</table>\n<p>1518 rows × 9 columns</p>\n</div>"
     },
     "metadata": {},
     "execution_count": 7
    }
   ],
   "source": [
    "metadata_bam_norep = pd.read_csv(\"data/evidences/encode/human/metadata_bam.no_rep.tsv\", sep=\"\\t\")\n",
    "metadata_bam_norep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I extract uniq `Experiment target` and count with which frequency the TF is under study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tail -n +2 data/evidences/encode/human/metadata_bam.tsv | cut -f 7 | cut -d '-' -f 1 | sort -u > data/evidences/encode/human/tf.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in `cat data/evidences/encode/human/tf.id`; do paste <(echo $i) <(grep -c $i data/evidences/encode/human/metadata_bam.no_rep.tsv); done | sort -grk 2 > data/evidences/encode/human/tf.freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TFs:\t489\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp target</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTCF</td>\n",
       "      <td>308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EP300</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JUN</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZNF2</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR3C1</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ZNF3</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FOS</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>REST</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CEBPB</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MYC</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ZNF7</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   exp target  freq\n",
       "0        CTCF   308\n",
       "1       EP300    50\n",
       "2         JUN    42\n",
       "3        ZNF2    41\n",
       "4       NR3C1    28\n",
       "5        ZNF3    26\n",
       "6         FOS    26\n",
       "7        REST    22\n",
       "8       CEBPB    22\n",
       "9         MYC    18\n",
       "10       ZNF7    17"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_freq = pd.read_csv(\"data/evidences/encode/human/tf.freq\", sep=\"\\t\", names=['exp target', 'freq'], header=None)\n",
    "print('Number of TFs:\\t%i' %len(tf_freq))\n",
    "tf_freq.loc[0:10,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I download the gene annotation file from **www.encodeproject.org/files**. Extract only protein coding genes and get name of the gene and gene's coordinates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wget -O data/evidences/encode/human/GRCh38_GENCODE_V29.gtf.gz https://www.encodeproject.org/files/ENCFF159KBI/@@download/ENCFF159KBI.gtf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zcat < data/evidences/encode/human/GRCh38_GENCODE_V29.gtf.gz \\\n",
    "    | awk 'BEGIN{FS=\"\\t\"}{split($9,a,\";\"); if(($3~\"gene\") && (a[2]~\"protein_coding\")) print a[3]\"\\t\"$1\"\\t\"$7\"\\t\"$4\"\\t\"$5}' \\\n",
    "    | sed 's/gene_name \"//; s/\"//g' | grep -v random | grep -v chrUn > data/evidences/encode/human/GRCh38_GENCODE_V29.table.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pol</th>\n",
       "      <th>sta</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>450703</td>\n",
       "      <td>451697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>923928</td>\n",
       "      <td>944581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>944204</td>\n",
       "      <td>959309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19935</th>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19936</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19937</th>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19938</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>chrM</td>\n",
       "      <td>-</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19939</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19940 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene   chr pol     sta     end\n",
       "0         OR4F5  chr1   +   65419   71585\n",
       "1        OR4F29  chr1   -  450703  451697\n",
       "2        OR4F16  chr1   -  685679  686673\n",
       "3        SAMD11  chr1   +  923928  944581\n",
       "4         NOC2L  chr1   -  944204  959309\n",
       "...         ...   ...  ..     ...     ...\n",
       "19935   MT-ND4L  chrM   +   10470   10766\n",
       "19936    MT-ND4  chrM   +   10760   12137\n",
       "19937    MT-ND5  chrM   +   12337   14148\n",
       "19938    MT-ND6  chrM   -   14149   14673\n",
       "19939    MT-CYB  chrM   +   14747   15887\n",
       "\n",
       "[19940 rows x 5 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch38_geneanno = pd.read_csv(\"data/evidences/encode/human/GRCh38_GENCODE_V29.table.gtf\", sep=\"\\t\", header=None, names=['gene', 'chr', 'pol', 'sta', 'end'])\n",
    "grch38_geneanno"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I modify the `GRCh38_GENCODE_V29.table.gtf` chromosomes' starts and ends to finally work with TIP method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "awk 'BEGIN{FS=\"\\t\"} \\\n",
    "    {if($3==\"+\") {start=($4-10000); end=($4+10000); print $1\"\\t\"$2\"\\t\"$3\"\\t\"start\"\\t\"end;} \\\n",
    "    else if($3==\"-\") {start=($5+10000); end=($5-10000); print $1\"\\t\"$2\"\\t\"$3\"\\t\"end\"\\t\"start;}}'\\\n",
    "    data/evidences/encode/human/GRCh38_GENCODE_V29.table.gtf > data/evidences/encode/human/GRCh38_GENCODE_V29.TIP20000.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>pol</th>\n",
       "      <th>sta</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>55419</td>\n",
       "      <td>75419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>441697</td>\n",
       "      <td>461697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>676673</td>\n",
       "      <td>696673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>913928</td>\n",
       "      <td>933928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>949309</td>\n",
       "      <td>969309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19935</th>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>470</td>\n",
       "      <td>20470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19936</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>760</td>\n",
       "      <td>20760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19937</th>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>2337</td>\n",
       "      <td>22337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19938</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>chrM</td>\n",
       "      <td>-</td>\n",
       "      <td>4673</td>\n",
       "      <td>24673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19939</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>chrM</td>\n",
       "      <td>+</td>\n",
       "      <td>4747</td>\n",
       "      <td>24747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19940 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene   chr pol     sta     end\n",
       "0         OR4F5  chr1   +   55419   75419\n",
       "1        OR4F29  chr1   -  441697  461697\n",
       "2        OR4F16  chr1   -  676673  696673\n",
       "3        SAMD11  chr1   +  913928  933928\n",
       "4         NOC2L  chr1   -  949309  969309\n",
       "...         ...   ...  ..     ...     ...\n",
       "19935   MT-ND4L  chrM   +     470   20470\n",
       "19936    MT-ND4  chrM   +     760   20760\n",
       "19937    MT-ND5  chrM   +    2337   22337\n",
       "19938    MT-ND6  chrM   -    4673   24673\n",
       "19939    MT-CYB  chrM   +    4747   24747\n",
       "\n",
       "[19940 rows x 5 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch38_geneanno = pd.read_csv(\"data/evidences/encode/human/GRCh38_GENCODE_V29.TIP20000.gtf\", sep=\"\\t\", header=None, names=['gene', 'chr', 'pol', 'sta', 'end'])\n",
    "grch38_geneanno"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RegNetwork\n",
    "\n",
    "**RegNetwork** is a database that stores knowledge-based genome-wide regulatory networks (Liu et al., 2015). So far it has been built for organisms as human and mouse by integrating 25 data sources. A comprehensive set of interplays among TFs, miRNAs and target genes are collected and reorganized for public access. Specifically, the mouse regulatory network contains 20738 nodes and 323636 edges, consisting of 1328 TFs, 1290 miRNAs and 18120 target genes. \n",
    "It should be noted that, besides the experimentally observed or discovered regulatory relationships that have been curated in public databases such as TRED and KEGG, RegNetwork integrates predictions based on TF binding site (TFBS) information for TF-gene and considers protein-protein interactions (PPIs) from HPRD, BioGrid, IntAct, KEGG and STRING as putative gene regulations when PPI pairs contain at least one TF (self-regulations are also identified in this case when TF-TF interactions is observed). It is a fortune that a degree of confidence for each of the regulatory interactions, by using a three-level labelling approach (i.e. a high, medium or low confidence), has been added. More specifically, the experiment-validated regulations are tagged with the label high confidence; the predictions made by only one algorithm/method are tagged with low confidence; and the rest are tagged with medium confidence. Although these confidence levels are not reported in the downloadable files from the download section of the website, one can use their search engine to filter out predictions and what has low or medium confidence. When one retrieves only the experiment-validated regulations with label high confidence, the coverage of the database decreases a lot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator_symbol</th>\n",
       "      <th>regulator_id</th>\n",
       "      <th>target_symbol</th>\n",
       "      <th>target_id</th>\n",
       "      <th>database</th>\n",
       "      <th>evidence</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>SHC3</td>\n",
       "      <td>53358</td>\n",
       "      <td>kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>STAT5B</td>\n",
       "      <td>6777</td>\n",
       "      <td>kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>CBLB</td>\n",
       "      <td>868</td>\n",
       "      <td>kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>CBLC</td>\n",
       "      <td>23624</td>\n",
       "      <td>kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>25</td>\n",
       "      <td>CD55</td>\n",
       "      <td>1604</td>\n",
       "      <td>kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9721</th>\n",
       "      <td>ZIC2</td>\n",
       "      <td>7546</td>\n",
       "      <td>GLI2</td>\n",
       "      <td>2736</td>\n",
       "      <td>hprd,kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9722</th>\n",
       "      <td>ZIC2</td>\n",
       "      <td>7546</td>\n",
       "      <td>GLI1</td>\n",
       "      <td>2735</td>\n",
       "      <td>hprd,kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9723</th>\n",
       "      <td>ZIC2</td>\n",
       "      <td>7546</td>\n",
       "      <td>GLI3</td>\n",
       "      <td>2737</td>\n",
       "      <td>hprd,kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9724</th>\n",
       "      <td>ZNF274</td>\n",
       "      <td>10782</td>\n",
       "      <td>NGFR</td>\n",
       "      <td>4804</td>\n",
       "      <td>hprd,kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9725</th>\n",
       "      <td>ZNF274</td>\n",
       "      <td>10782</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>7189</td>\n",
       "      <td>kegg</td>\n",
       "      <td>Experimental</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9726 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator_symbol  regulator_id target_symbol target_id   database  \\\n",
       "0                ABL1            25          SHC3     53358       kegg   \n",
       "1                ABL1            25        STAT5B      6777       kegg   \n",
       "2                ABL1            25          CBLB       868       kegg   \n",
       "3                ABL1            25          CBLC     23624       kegg   \n",
       "4                ABL1            25          CD55      1604       kegg   \n",
       "...               ...           ...           ...       ...        ...   \n",
       "9721             ZIC2          7546          GLI2      2736  hprd,kegg   \n",
       "9722             ZIC2          7546          GLI1      2735  hprd,kegg   \n",
       "9723             ZIC2          7546          GLI3      2737  hprd,kegg   \n",
       "9724           ZNF274         10782          NGFR      4804  hprd,kegg   \n",
       "9725           ZNF274         10782         TRAF6      7189       kegg   \n",
       "\n",
       "          evidence confidence  \n",
       "0     Experimental       High  \n",
       "1     Experimental       High  \n",
       "2     Experimental       High  \n",
       "3     Experimental       High  \n",
       "4     Experimental       High  \n",
       "...            ...        ...  \n",
       "9721  Experimental       High  \n",
       "9722  Experimental       High  \n",
       "9723  Experimental       High  \n",
       "9724  Experimental       High  \n",
       "9725  Experimental       High  \n",
       "\n",
       "[9726 rows x 7 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "report = pd.read_csv(\"data/db/regnet/human/export_Tue_Apr_07_20_14_32_UTC_2020.csv\", sep=\",\")\n",
    "report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Activation': [],\n",
       " 'Repression': [],\n",
       " 'Unknown': ['NGFR', 'TRAF6', 'NGFR', 'TRAF6', 'NGFR', 'TRAF6']}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"data/db/regnet/human/regnet_db.pkl\", 'rb') as f1:\n",
    "    regnet_db = pickle.load(f1)\n",
    "regnet_db['ZNF274']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TRRUSTv2\n",
    "**TRRUST** (Transcriptional Regulatory Relationships Unraveled by Sentence-based Text-mining) is a database of reference TF–target regulatory interactions identified via the manual curation of Medline abstracts (Han et al., 2015). Sentence-based text mining and prioritization of pre-candidate sentences is conducted to reduce the cost-effective literature curation by differentiating sentences for their likelihood of containing information regarding TF–gene interactions.\n",
    "\n",
    "The current version of TRRUSTv2 contains 6490 TF–target interactions for 827 mouse's TFs.  Most importantly, a majority of the interactions have annotations for mode-of-regulation (i.e., activation or repression).  If a sentence suggested a mode-of-regulation, i.e., either activation or repression, for the given TF-gene regulatory interaction, the authors collected and deposited the information along with the given TF-gene interaction in the database. If there was no indication in either the given sentence or its associated abstract, then the term unknown for the mode-of-regulation of the TF-gene regulatory interaction was assigned. Currently 8972 (59.8%) TF-gene interactions, over both human and mouse 14996 regulatory relationships, are known for mode of regulation.\n",
    "\n",
    "This exclusive to TRRUSTv2 could be useful if one would like to add annotation to new directed FunCoup links being mode-of-regulation important in interpreting the phenotype effects of TF dysregulation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator_symbol</th>\n",
       "      <th>target_symbol</th>\n",
       "      <th>mode_of_regulation</th>\n",
       "      <th>PMID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AATF</td>\n",
       "      <td>BAX</td>\n",
       "      <td>Repression</td>\n",
       "      <td>22909821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AATF</td>\n",
       "      <td>CDKN1A</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>17157788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AATF</td>\n",
       "      <td>KLK3</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>23146908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AATF</td>\n",
       "      <td>MYC</td>\n",
       "      <td>Activation</td>\n",
       "      <td>20549547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AATF</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>17157788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9391</th>\n",
       "      <td>ZNF76</td>\n",
       "      <td>CDKN1A</td>\n",
       "      <td>Repression</td>\n",
       "      <td>15280358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9392</th>\n",
       "      <td>ZNF76</td>\n",
       "      <td>PCYT1A</td>\n",
       "      <td>Activation</td>\n",
       "      <td>14702349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9393</th>\n",
       "      <td>ZNF76</td>\n",
       "      <td>TALDO1</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>14702349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9394</th>\n",
       "      <td>ZNRD1</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>Activation</td>\n",
       "      <td>16373708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9395</th>\n",
       "      <td>ZNRD1</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>Activation</td>\n",
       "      <td>16373708;16609701</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9396 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator_symbol target_symbol mode_of_regulation               PMID\n",
       "0                AATF           BAX         Repression           22909821\n",
       "1                AATF        CDKN1A            Unknown           17157788\n",
       "2                AATF          KLK3            Unknown           23146908\n",
       "3                AATF           MYC         Activation           20549547\n",
       "4                AATF          TP53            Unknown           17157788\n",
       "...               ...           ...                ...                ...\n",
       "9391            ZNF76        CDKN1A         Repression           15280358\n",
       "9392            ZNF76        PCYT1A         Activation           14702349\n",
       "9393            ZNF76        TALDO1            Unknown           14702349\n",
       "9394            ZNRD1         ABCB1         Activation           16373708\n",
       "9395            ZNRD1          BCL2         Activation  16373708;16609701\n",
       "\n",
       "[9396 rows x 4 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "report = pd.read_csv(\"data/db/trrust/human/trrust_rawdata.human.tsv\", sep=\"\\t\", header=None, names=['regulator_symbol', 'target_symbol', 'mode_of_regulation', 'PMID'])\n",
    "report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Activation': [],\n",
       " 'Repression': [],\n",
       " 'Unknown': ['BAX', 'CDKN1A', 'KLK3', 'MYC', 'TP53']}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"data/db/trrust/human/trrust_db.pkl\", 'rb') as f1:\n",
    "    trrust_db = pickle.load(f1)\n",
    "trrust_db['AATF']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I get from `https://ftp.ncbi.nlm.nih.gov/genomes/` the information about name and lenghts of chromosomes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <td>chr1</td>\n",
       "      <td>chr2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>chr4</td>\n",
       "      <td>chr5</td>\n",
       "      <td>chr6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>chr8</td>\n",
       "      <td>chr9</td>\n",
       "      <td>chr10</td>\n",
       "      <td>...</td>\n",
       "      <td>chr16</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr18</td>\n",
       "      <td>chr19</td>\n",
       "      <td>chr20</td>\n",
       "      <td>chr21</td>\n",
       "      <td>chr22</td>\n",
       "      <td>chrX</td>\n",
       "      <td>chrY</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>length</th>\n",
       "      <td>248956422</td>\n",
       "      <td>242193529</td>\n",
       "      <td>198295559</td>\n",
       "      <td>190214555</td>\n",
       "      <td>181538259</td>\n",
       "      <td>170805979</td>\n",
       "      <td>159345973</td>\n",
       "      <td>145138636</td>\n",
       "      <td>138394717</td>\n",
       "      <td>133797422</td>\n",
       "      <td>...</td>\n",
       "      <td>90338345</td>\n",
       "      <td>83257441</td>\n",
       "      <td>80373285</td>\n",
       "      <td>58617616</td>\n",
       "      <td>64444167</td>\n",
       "      <td>46709983</td>\n",
       "      <td>50818468</td>\n",
       "      <td>156040895</td>\n",
       "      <td>57227415</td>\n",
       "      <td>16569</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0          1          2          3          4          5   \\\n",
       "chr          chr1       chr2       chr3       chr4       chr5       chr6   \n",
       "length  248956422  242193529  198295559  190214555  181538259  170805979   \n",
       "\n",
       "               6          7          8          9   ...        15        16  \\\n",
       "chr          chr7       chr8       chr9      chr10  ...     chr16     chr17   \n",
       "length  159345973  145138636  138394717  133797422  ...  90338345  83257441   \n",
       "\n",
       "              17        18        19        20        21         22        23  \\\n",
       "chr        chr18     chr19     chr20     chr21     chr22       chrX      chrY   \n",
       "length  80373285  58617616  64444167  46709983  50818468  156040895  57227415   \n",
       "\n",
       "           24  \n",
       "chr      chrM  \n",
       "length  16569  \n",
       "\n",
       "[2 rows x 25 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if organism == 'human':\n",
    "    chr_nam = ['chr' + str(i) for i in range(1,23)] + ['chrX', 'chrY', 'chrM']\n",
    "    if genome == 'GRCh38': \n",
    "        chr_len = [ 248956422, 242193529, 198295559, 190214555, 181538259, 170805979, 159345973, \n",
    "                    145138636, 138394717, 133797422, 135086622, 133275309, 114364328, 107043718, \n",
    "                    101991189, 90338345, 83257441, 80373285, 58617616, 64444167, 46709983, 50818468, \n",
    "                    156040895, 57227415, 16569 ]\n",
    "\n",
    "dictio = {}\n",
    "for i in range(len(chr_nam)):\n",
    "    dictio[chr_nam[i]] = dictio.get(chr_nam[i], chr_len[i])\n",
    "pd.DataFrame(dictio.items(), columns=['chr','length']).T"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}